### PR TITLE
Improve sass grammar

### PR DESCRIPTION
**Pull request by idbartosz**
Adds support for missing at-rules in sass.cson (#171)

improve-sass-grammar

### CODE REVIEW

**Comment by 50Wliu** [comment]
This will need specs.

**Comment by idbartosz** [comment]
Due to unit fail I've discarded property-value change. Basically it was a small improvement in regex to not act greedy with white spaces in variables.

property-fvalue-fix

**Comment by idbartosz** [comment]
:white_check_mark: Success: The Travis CI build passed

**Comment by idbartosz** [comment]
Added specs.

File: grammars/sass.cson [hunk 1]
```diff
@@ -338,6 +529,7 @@
     ]
   'property-value':
     'begin': '(:)?\\s?+'
+    # 'begin': '(:)?(?=\\S+)' # discard due unit fail
```
Comment (50Wliu) [inline grammars/sass.cson]: Why was this added?

File: grammars/sass.cson [hunk 2]
```diff
@@ -373,7 +561,7 @@
...
...ase-out|ease|ellipsis|fill|fixed|grid|groove|hand|help|hidden|horizontal|ideograph-alpha|ideograph-numeric|ideograph-parenthesis|ideograph-space|inactive|inherit|inline-block|inline-flex|inline-grid|inline-table|inline|inset|inside|inter-ideograph|inter-word|italic|justify|keep-all|left|lighter|line-edge|line-through|linear|line|list-item|loose|lower-alpha|lower-roman|lowercase|lr-tb|ltr|manipulation|medium|middle|move|n-resize|ne-resize|newspaper|no-drop|no-repeat|nw-resize|none|normal|not-allowed|nowrap|oblique|outset|outside|overline|padding-box|pan-(x|y|left|right|up|down)|pointer|pre-wrap|pre-line|pre|preserve-3d|progress|relative|repeat-x|repeat-y|repeat|right|ridge|round|row-resize|rtl|ruby-base-container|ruby-base|ruby-text-container|ruby-text|ruby|s-resize|scale-down|scroll|se-resize|separate|small-caps|solid|square|static|step-end|step-start|strict|super|sw-resize|table-caption|table-cell|table-column-group|table-column|table-footer-group|table-header-group|table-row-group|table-row|table|tb-rl|text-bottom|text-top|text|thick|thin|top|transparent|underline|upper-alpha|upper-roman|uppercase|vertical-ideographic|vertical-text|visible|w-resize|wait|whitespace|true|false)\\b'
```
Comment (50Wliu) [inline grammars/sass.cson]: Keep this alphabetical please

File: spec/sass-spec.coffee [hunk 3]
```diff
@@ -144,3 +144,127 @@ describe 'SASS grammar', ->
       expect(tokens[0][1]).toEqual value: ' hi1', scopes: ['source.sass', 'comment.line.sass']
       expect(tokens[1][0]).toEqual value: '  hi2', scopes: ['source.sass', 'comment.line.sass']
       expect(tokens[2][0]).not.toEqual value: 'hi3', scopes: ['source.sass', 'comment.line.sass']
+
+  describe 'correctly tokenizes at-rules and directives with proper syntax', ->
```
Comment (50Wliu) [inline spec/sass-spec.coffee]: `describe 'at-rules'`

File: spec/sass-spec.coffee [hunk 4]
```diff
@@ -144,3 +144,127 @@ describe 'SASS grammar', ->
...
+    it '@function', ->
```
Comment (50Wliu) [inline spec/sass-spec.coffee]: `it 'tokenizes @function'`

Repeat this for all the other ones.

File: spec/sass-spec.coffee [hunk 5]
```diff
@@ -144,3 +144,127 @@ describe 'SASS grammar', ->
...
+      expect(tokens[1]).toEqual value: 'function', scopes: ['source.sass', 'meta.at-rule.function.sass', 'keyword.control.at-rule.function.sass']
+      expect(tokens[3]).toEqual value: 'function_name', scopes: ['source.sass', 'meta.at-rule.function.sass', 'support.function.misc.sass']
+
+    it '@return', ->
+      {tokens} = grammar.tokenizeLine '@return \'border\' + \' \' + \'1px solid pink\''
+      expect(tokens[0]).toEqual value: '@', scopes: ['source.sass', 'meta.at-rule.return.sass', 'keyword.control.return.sass', 'punctuation.definition.entity.sass']
+      expect(tokens[1]).toEqual value: 'return', scopes: ['source.sass', 'meta.at-rule.return.sass', 'keyword.control.return.sass']
+
+    it '@if', ->
+      {tokens} = grammar.tokenizeLine '@if $var == true'
```
Comment (50Wliu) [inline spec/sass-spec.coffee]: Does `$var` get matched as a variable?

File: spec/sass-spec.coffee [hunk 6]
```diff
@@ -144,3 +144,127 @@ describe 'SASS grammar', ->
...
+      expect(tokens[0]).toEqual value: '@', scopes: ['source.sass', 'meta.at-rule.else.sass', 'keyword.control.else.sass', 'punctuation.definition.entity.sass']
+      expect(tokens[1]).toEqual value: 'else if ', scopes: ['source.sass', 'meta.at-rule.else.sass', 'keyword.control.else.sass']
+
+    it '@while', ->
+      {tokens} = grammar.tokenizeLine '@while 1'
+      expect(tokens[0]).toEqual value: '@', scopes: ['source.sass', 'meta.at-rule.while.sass', 'keyword.control.while.sass', 'punctuation.definition.entity.sass']
+      expect(tokens[1]).toEqual value: 'while', scopes: ['source.sass', 'meta.at-rule.while.sass', 'keyword.control.while.sass']
+
+    it '@for', ->
+      {tokens} = grammar.tokenizeLine '@for $i from 1 through 100'
```
Comment (50Wliu) [inline spec/sass-spec.coffee]: Similarly, what about the `from` and the `through`?  Are those tokenized correctly?  Some more examples are below.

File: grammars/sass.cson (hunk 1 above)
Comment (idbartosz) [inline grammars/sass.cson]: It's a future fix and I've written about it in 3rd comment.

File: grammars/sass.cson (hunk 2 above)
Comment (idbartosz) [inline grammars/sass.cson]: Ok.

File: spec/sass-spec.coffee (hunk 3 above)
Comment (idbartosz) [inline spec/sass-spec.coffee]: Ok, but @if, @for and others are Sass control directives.

File: spec/sass-spec.coffee (hunk 4 above)
Comment (idbartosz) [inline spec/sass-spec.coffee]: Ok.

File: spec/sass-spec.coffee (hunk 5 above)
Comment (idbartosz) [inline spec/sass-spec.coffee]: Sure, it even detects when the statment isn't correct but I'll add tests.

File: spec/sass-spec.coffee (hunk 6 above)
Comment (idbartosz) [inline spec/sass-spec.coffee]: You got that tests in line 263, 264, 267 and 270 because those are parts of operators.

Comment (50Wliu) [inline spec/sass-spec.coffee]: Ahaha, I see those now.  Thanks.

File: spec/sass-spec.coffee (hunk 3 above)
Comment (50Wliu) [inline spec/sass-spec.coffee]: How about `describe 'at-rules and directives` then?

Comment (idbartosz) [inline spec/sass-spec.coffee]: Works for me, I thought BDD is more desctriptive :-)

**Comment by idbartosz** [comment]
@50Wliu
I hope arrangements were fulfilled

File: grammars/sass.cson (hunk 1 above)
Comment (50Wliu) [inline grammars/sass.cson]: If it's for a future fix, please remove this.

Comment (idbartosz) [inline grammars/sass.cson]: Done.

**Comment by 50Wliu** [comment]
Thanks!
